cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/resource/KafkaBrokerConfigurationDiff.java

```diff
@@ -0,0 +1,201 @@
+        map.entrySet().forEach(entry -> {
+            if (!IGNORABLE_PROPERTIES.matcher(entry.getKey()).matches() && entry.getValue().contains("${" + placeholder + "}")) {
+                entry.setValue(entry.getValue().replaceAll("\\$\\{" + placeholder + "\\}", value));
+            }
+        });
+    }
+
+    private KafkaConfiguration computeDiff() {
+        HashMap<String, String> desiredMap = configMap2Map();
+        Map<String, String> currentMap = new HashMap<>();
+
+        currentEntries.stream().forEach(e -> {
+            currentMap.put(e.name(), e.value());
+        });
+
+        fillPlaceholderValue(desiredMap, "STRIMZI_BROKER_ID", Integer.toString(brokerId));
+
+        Map<String, String> diff = getMapDiff(currentMap, desiredMap);
+
+        String diffString = diff.toString();
+        diffString = diffString.substring(1, diffString.length() - 1).replace(", ", "\n");
+
+        return KafkaConfiguration.unvalidated(diffString);
+    }
+
+    public KafkaConfiguration getDiff() {
+        return this.diff;
+    }
+
+    @SuppressWarnings("checkstyle:Regexp")
+    private Map<String, String> getMapDiff(Map<String, String> current, Map<String, String> desired) {
+        Map<String, String> difference = new HashMap<>();
+        difference.putAll(desired);
+        desired.forEach((k, v) -> {
+            if (IGNORABLE_PROPERTIES.matcher(k).matches() || (current.get(k) != null && current.get(k).toLowerCase(Locale.ENGLISH).equals(v.toLowerCase(Locale.ENGLISH)))) {
+                difference.remove(k);
+            } else {
+                log.info("{} differs in '{}' ---> '{}'", k, current.get(k), v);
+            }
+        });
+
+        current.entrySet().forEach(entry -> {
+            // some value was set to non-default value, then the entry was removed from desired -> we want to use default value
+            if (!desired.keySet().contains(entry.getKey()) && !isDesiredPropertyDefaultValue(entry.getKey(), entry.getValue())) {
+                if (!IGNORABLE_PROPERTIES.matcher(entry.getKey()).matches()) {
+                    String defVal = KafkaConfiguration.getDefaultValueOfProperty(entry.getKey(), kafkaVersion) == null ? "null" : KafkaConfiguration.getDefaultValueOfProperty(entry.getKey(), kafkaVersion).toString();
+                    log.info("{} had value {} and was removed from desired. Setting {}", entry.getKey(), entry.getValue(), defVal);
+                    difference.put(entry.getKey(), defVal);
+                    deletedEntries.add(entry.getKey());
+                }
+            }
+        });
+
+        return difference;
+    }
+
+    public boolean isDesiredPropertyDefaultValue(String key, String value) {
+        Optional<ConfigEntry> entry = currentEntries.stream().filter(configEntry -> configEntry.name().equals(key)).findFirst();
+        if (entry.isPresent()) {
+            return KafkaConfiguration.isValueOfPropertyDefault(entry.get().name(), value, kafkaVersion);
```

Comment:
Do you know about `ConfigEntry.ConfigSource`?

Reply:
Yes, I was looking at this, but then algorithm is dependent on which version of kafka does strimzi use, does not it? If we want to use a different version than it was built with, it could make problems, or?

Reply:
When you `describeConfigs()` you're get back (from the _broker_) each config and where it came from, which is what you need to know isn't it? It doesn't depend on the version of the AdminClient used by the CO. 

Also, bear in mind that the broker might be configured with a value which == the default (e.g. `DEFAULT_CONFIG`), but the value is not coming from the default (e.g. it might be `DYNAMIC_DEFAULT_BROKER_CONFIG` or `STATIC_BROKER_CONFIG`). IDK how that affects your algorithm. 

Also bear in mind that when changing a cluster-scoped config it's "good practice" to just set it for an indiviual broker (i.e. `DYNAMIC_BROKER_CONFIG) and verify that everything is OK before applying it to the whole cluster (i.e. as `DYNAMIC_DEFAULT_BROKER_CONFIG`). See https://kafka.apache.org/documentation.html#dynamicbrokerconfigs. IDK if you want to tackle that in this PR or not, but thought it worth mentioning.